power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -200,124 +354,209 @@ class Deserializer {
+            throw SerializationError{"Key type not found!\n"};
+        }
+        if (!has_is_batch) {
+            throw SerializationError{"Key is_batch not found!\n"};
+        }
+        if (!has_attributes) {
+            throw SerializationError{"Key attributes not found!\n"};
         }
-        msgpack::object const& obj = as_map(map).ptr[idx].val;
-        if (obj.type != type) {
-            throw SerializationError{"Wrong data type for key " + std::string(key) +
-                                     " in the root level dictionary!\n"};
+        if (!has_data) {
+            throw SerializationError{"Key data not found!\n"};
         }
-        return obj;
+
+        WritableDatasetHandler handler{is_batch_, batch_size, dataset};
+        count_data(handler, data_counts);
+        parse_predefined_attributes(handler.dataset(), attributes);
+        return handler;
     }
 
-    static Idx find_key_from_map(msgpack::object const& map, std::string_view key) {
-        auto const kv_map = map.as<MapSpan>();
-        auto const found = std::ranges::find_if(kv_map, [key](auto const& x) { return key_to_string(x) == key; });
-        if (found == kv_map.end()) {
-            return -1;
+    AttributeByteMeta read_predefined_attributes() {
+        AttributeByteMeta attributes;
+        Idx n_components = parse_map_array<true, false, true>().size;
+        while (n_components-- != 0) {
+            component_key_ = parse_string();
+            attributes.push_back({component_key_, {}});
+            auto& attributes_per_component = attributes.back().second;
+            Idx const n_attributes_per_component = parse_map_array<false, true, true>().size;
+            for (element_number_ = 0; element_number_ != n_attributes_per_component; ++element_number_) {
+                attributes_per_component.push_back(parse_string());
+            }
+            element_number_ = -1;
         }
-        return std::distance(kv_map.begin(), found);
+        component_key_ = {};
+        return attributes;
     }
 
-    void read_predefined_attributes() {
-        for (auto const& kv : get_value_from_root("attributes", msgpack_map).as<MapSpan>()) {
-            component_key_ = key_to_string(kv);
-            MetaComponent const& component = dataset_handler_.dataset().get_component(component_key_);
-            attributes_[component.name] = read_component_attributes(component, kv.val);
+    void parse_predefined_attributes(MetaDataset const& dataset, AttributeByteMeta const& attributes) {
+        root_key_ = "attributes";
+        for (auto const& single_component : attributes) {
+            component_key_ = single_component.first;
+            MetaComponent const* const component = &dataset.get_component(component_key_);
+            std::vector<MetaAttribute const*> attributes_per_component;
+            for (element_number_ = 0; element_number_ != static_cast<Idx>(single_component.second.size());
+                 ++element_number_) {
```

Comment:
since we now have `<ranges>`, can you try to see whether this works?

```suggestion
            for (element_number_ : std::views::iota(0, static_cast<Idx>(single_component.second.size())) {
```

of not, please use `boost::counting_iterator`